File: neetcode all and misc questions/Heaps/last_stone_weight.py
"""
https://leetcode.com/problems/last-stone-weight/
"""
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        """
        https://neetcode.io/problems/last-stone-weight
        Use a max heap 
        continue popping off max heap twice at a time until there's at least one element in the max heap left
        O(NLogN) + O(MLogN), where M is the amount of times heappush() is called
        """
        import heapq
        # use negative numbers to create a max heap
        maxHeap = [-stones[i] for i in range(len(stones))]
        heapq.heapify(maxHeap)
        # there has to be at least two stones
        while len(maxHeap) > 1:
            # convert back to positive numbers
            stone1 = -1 * heapq.heappop(maxHeap)
            stone2 = -1 * heapq.heappop(maxHeap)
            if stone1 > stone2 or stone2 > stone1:
                # get the weight difference and multiply by -1
                # to put back onto the max heap
                newStone = -1 * abs(stone1-stone2)
                heapq.heappush(maxHeap, newStone)
        return -1 * maxHeap[0] if len(maxHeap) == 1 else 0

class Solution:
    """
    Max Heap Solution
    """
    def lastStoneWeight2(self, stones: List[int]) -> int:
        import heapq
        maxHeap = [-stones[i] for i in range(len(stones))]
        heapq.heapify(maxHeap)
        # take the first two elements off the top of the heap
        while (len(maxHeap) > 1 and maxHeap[0] != 0):
            # by doing element - element, this can be 0 if the elements have the same value,
            # which is why we have the top condition mapHeap[0] != 0
            heapq.heappush(maxHeap, heapq.heappop(maxHeap)-heapq.heappop(maxHeap))
        # return - to turn the number back into positive
        return -maxHeap[0]



    """
    Brute Force
    while loop, while there are at least 2 stones,
    continue the computations, and then re-sort
    after each computation
    O(N * NLogN)
    """
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones.sort()
        if len(stones) == 1:
            return stones[-1]
        while (len(stones) > 1):
            y = stones[-1]
            x = stones[-2]
            if x == y:
                # destroy both
                stones = stones[:-2]
            else:
                # x is destroyed
                # y is recalculated
                newY = y - x
                stones = stones[:-2] + [newY]
            stones.sort()
        return stones[-1] if len(stones) > 0 else 0

File: node_js_ds_and_algos_implementations/heaps/last_stone_weight.js
class Solution {
    /**
     * @param {number[]} stones
     * @return {number}
     */
    lastStoneWeight(stones) {
        /* 
        Revisited 9/22/2026
        max heap
        pop two from the front and applying the logic
        pushing back in if the stone isn't destroyed
        */

        let maxHeap = new MaxPriorityQueue()
        for (let i = 0; i < stones.length; ++i){
            maxHeap.enqueue(stones[i])
        }
        while (maxHeap.size() > 1){
            let first = maxHeap.dequeue()
            let second = maxHeap.dequeue()
            if (first !== second){
                let difference = Math.abs(first-second)
                maxHeap.enqueue(difference)
            }
        }
        return maxHeap.size() > 0 ? maxHeap.front() : 0
    }
}

/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function(stones) {
    /*
    convert array to max heap to maintain sorted order of the stone weights
    
    while max heap size > 1
        pop the first two values off the heap
        apply the logic to figure out the new weight
        push the new weight to the heap if valid

    Time complexity:
    O(NLogK) (where K is the height of the heap, and N is the number of stones)
    Each logK is referring to the complexity needed to push and pop a node.
    Space complexity:
    O(N) for the heap
    */
    let heap = new MaxPriorityQueue()
    for (let stone of stones){
        heap.enqueue(stone)
    }
    while (heap.size() > 1){
        let { element: s1 } = heap.dequeue()
        let { element: s2 } = heap.dequeue()
        if (s1 !== s2){
            heap.enqueue(s1-s2)
        }
    }
    return heap.size() === 0 ? 0 : heap.front().element
    
    
};